### ISSUE TITLE
loading message replaces the server-rendered code

### ISSUE THREAD

**Issue by codejamninja** [ReactTraining/react-router#6428]
Displaying a temporary loading message while routes are dynamically loading is great in theory, but horrible for server-side rendered isomorphic applications. The following describes what happens.

* The server renders the route and sends the generated html to the client
* The client renders the html
* The client loads the javascript bundle
* Client-side rendering takes over and rerenders the page
* While the route component is dynamically loading, a temporary **loading message replaces the server-rendered code** (THIS IS THE BUG)

In summary, while the client rendering takes over, the temporary loading message replaces the server-rendered code.

Does anyone have any suggestions for working around this?

https://github.com/ReactTraining/react-router/issues/4407

**Comment by timdorr** [comment]
There are many examples of this in the linked issue. Short answer: you need some way of pre-populating the state in the component from the page. Normally, this is something like window.__BOOTSTRAP_STATE, which is filled in by the server and used once. There are many ways to do it. It isn't a React Router problem, though.

**Comment by codejamninja** [comment]
Ok, but not having an asyncrous render function is a setback in react router. Can you please give me a good reason not to make the render function in react router asynchronous?
